separatedunit.cpp

```diff
@@ -0,0 +1,1286 @@
+/// specific phase.
+SeparatedUnit::SeparatedUnit(bitLenInt qBitCount, bitCapInt initState, Complex16 phaseFac)
+{
+    rand_generator_ptr[0] = std::default_random_engine();
+    randomSeed = std::time(0);
+    SetRandomSeed(randomSeed);
+    qubitCount = qBitCount;
+    maxQPower = 1 << qubitCount;
+
+    if (phaseFac == Complex16(-999.0, -999.0)) {
+        double angle = Rand() * 2.0 * M_PI;
+        phaseFac = Complex16(cos(angle), sin(angle));
+    }
+
+    bool setBit;
+    bitLenInt i;
+    std::unique_ptr<QbLookup[]> ql(new QbLookup[qBitCount]);
+    std::unique_ptr<bitLenInt[]> qil(new bitLenInt[qBitCount * qBitCount]());
+    qubitLookup = std::move(ql);
+    qubitInverseLookup = std::move(qil);
+    for (i = 0; i < qBitCount; i++) {
+        setBit = (initState & (1 << i)) > 0;
+        qubitLookup[i].cu = i;
+        qubitLookup[i].qb = 0;
+        qubitInverseLookup[i * qBitCount] = i;
+        coherentUnits.push_back(
+            std::shared_ptr<CoherentUnit>(new CoherentUnit(1, (setBit ? 1 : 0), phaseFac, rand_generator_ptr)));
+    }
+}
+
+/// Initialize a coherent unit with qBitCount number of bits, to initState unsigned integer permutation state, with a
+/// specific phase.
+SeparatedUnit::SeparatedUnit(bitLenInt qBitCount, bitCapInt initState)
+    : SeparatedUnit(qBitCount, initState, Complex16(-999.0, -999.0))
+{
+}
+
+/// Initialize a coherent unit with qBitCount number of bits, all to |0> state.
+SeparatedUnit::SeparatedUnit(bitLenInt qBitCount)
+    : SeparatedUnit(qBitCount, 0, Complex16(-999.0, -999.0))
+{
+}
+
+/// Initialize a coherent unit with qBitCount number of bits, all to |0> state, with a specific phase.
+SeparatedUnit::SeparatedUnit(bitLenInt qBitCount, Complex16 phaseFac)
+    : SeparatedUnit(qBitCount, 0, phaseFac)
+{
+}
+
+SeparatedUnit::SeparatedUnit(const SeparatedUnit& pqs)
+{
+    rand_generator_ptr[0] = std::default_random_engine();
+    randomSeed = std::time(0);
+    SetRandomSeed(randomSeed);
+    qubitCount = pqs.qubitCount;
+    maxQPower = 1 << qubitCount;
+
+    std::unique_ptr<QbLookup[]> ql(new QbLookup[qubitCount]);
+    std::copy(&(pqs.qubitLookup[0]), &(pqs.qubitLookup[0]) + qubitCount, &(ql[0]));
+    qubitLookup = std::move(ql);
+    std::unique_ptr<bitLenInt[]> qil(new bitLenInt[qubitCount * qubitCount]());
+    std::copy(&(pqs.qubitInverseLookup[0]), &(pqs.qubitInverseLookup[0]) + qubitCount, &(qil[0]));
+    qubitLookup = std::move(ql);
+
+    for (bitLenInt i = 0; i < pqs.coherentUnits.size(); i++) {
+        coherentUnits.push_back(
+            std::shared_ptr<CoherentUnit>(new CoherentUnit(*(pqs.coherentUnits[pqs.coherentUnits.size() - i - 1]))));
+    }
+}
+
+void SeparatedUnit::Cohere(CoherentUnit& toCopy)
+{
+    bitLenInt i, cuLen;
+    bitLenInt qubitsToAdd = toCopy.GetQubitCount();
+    std::unique_ptr<QbLookup[]> ql(new QbLookup[qubitCount + qubitsToAdd]);
+    std::unique_ptr<bitLenInt[]> qil(new bitLenInt[(qubitCount + qubitsToAdd) * (qubitCount + qubitsToAdd)]());
```

Comment:
The QIL maps from an arbitrary bit within an arbitrary CU to a particular qubit from the perspective of the SU.  Because SU may entangle qubit 1 and 3, and have both of them in the same CU as a result, bit 2 in that CU needs to be map-able for some algorithms back to SU:qubit 3.

This is a placeholder comment to come back to this later with an alternative proposal to allocating a n^2 array.